pkg/pushsync/pushsync.go

```diff
@@ -332,118 +329,101 @@ func (ps *PushSync) pushToClosest(ctx context.Context, ch swarm.Chunk, origin bo
 	defer span.Finish()
 	defer ps.skipList.PruneExpired()
 
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
 	var (
-		// limits "attempted" requests, see pushPeer when a request becomes attempted
-		allowedPushes = 1
-		// limits total requests, irregardless of "attempted"
-		allowedRetries = maxPeers
+		sentErrorsLeft   = 1
+		preemptiveTicker <-chan time.Time
+		includeSelf      = ps.includeSelf
+		skip             []swarm.Address
+		inflight         int
 	)
 
 	if origin {
-		allowedPushes = maxAttempts
+		ticker := time.NewTicker(preemptiveInterval)
+		defer ticker.Stop()
+		preemptiveTicker = ticker.C
+		sentErrorsLeft = maxPushErrors
 	}
 
-	var (
-		includeSelf = ps.includeSelf
-		skipPeers   = new(skippeers.List)
-	)
-
 	resultChan := make(chan receiptResult)
-	doneChan := make(chan struct{})
-	defer close(doneChan)
+	done := make(chan struct{})
+	defer close(done)
 
-	timer := time.NewTimer(0)
-	defer timer.Stop()
+	retryC := make(chan struct{})
+
+	retry := func() {
+		go func() {
+			select {
+			case retryC <- struct{}{}:
+			case <-done:
+			}
+		}()
+	}
+
+	retry()
 
 	// nextPeer attempts to lookup the next peer to push the chunk to, if there are overdrafted peers the boolean would signal a re-attempt
-	nextPeer := func() (peer swarm.Address, retry bool, err error) {
+	nextPeer := func() (peer swarm.Address, err error) {
 
-		fullSkipList := append(ps.skipList.ChunkSkipPeers(ch.Address()), skipPeers.All()...)
+		fullSkipList := append(skip, ps.skipList.ChunkPeers(ch.Address())...)
 
 		peer, err = ps.topologyDriver.ClosestPeer(ch.Address(), includeSelf, topology.Filter{Reachable: true}, fullSkipList...)
 		if err != nil {
-			// ClosestPeer can return ErrNotFound in case we are not connected to any peers
-			// in which case we should return immediately.
-			// if ErrWantSelf is returned, it means we are the closest peer.
 			if errors.Is(err, topology.ErrWantSelf) {
-
 				if !ps.warmedUp() {
-					return swarm.ZeroAddress, false, ErrWarmup
+					return swarm.ZeroAddress, ErrWarmup
 				}
 
-				if skipPeers.OverdraftListEmpty() { // no peers in skip list means we can be confident that we are the closest peer
-					// we don't act on ErrWantSelf unless there are no overdraft peers
-					if !ps.radiusChecker.IsWithinStorageRadius(ch.Address()) {
-						return swarm.ZeroAddress, false, ErrOutOfDepthStoring
-					}
-					ps.pushToNeighbourhood(ctx, fullSkipList, ch, origin, originAddr)
-					return swarm.ZeroAddress, false, err
+				if !ps.radiusChecker.IsWithinStorageRadius(ch.Address()) {
```

Comment:
This changes the logic quite a bit. We have removed the whole overdraft logic in order to make this skiplist generic. The point of the overdraft peers was that you can retry again as they could potentially be usable. But now we will no longer consider them.

Reply:
It was discussed a little more than a week back that after not finding a single case in mainnet of having to delay retries in the event of having overdrafted peers, supporting this extra logic is not needed, the retry mechanism is sufficient.